src/mcpdotnet/Configuration/McpServerBuilderExtensions.Tools.cs

```diff
@@ -16,19 +19,54 @@ namespace McpDotNet;
 /// </summary>
 public static partial class McpServerBuilderExtensions
 {
+    internal const string RequiresUnreferencedCodeMessage = "This method might look up types that cannot be statically analyzed.";
```

Comment:
This component needs a bit more work: each method needs to also accept an optional `JsonSerializerOptions` parameter similar to how `AIFunctionFactory` works in Microsoft.Extensions.AI.

The overload accepting `params Type[] toolTypes` cannot be made to work well with the linker since the `DynamicallyAccessesMembers` attribute only works with `Type` or `string` parameters. My recommendation would be to remove the overload altogether in favor of a chaining pattern like so:

```C#
builder.WithTool(typeof(Tool1)).WithTool(typeof(Tool2));
```

This pattern would ensure trim compatibility for applications.

I held short of making any changes like that, but would be interested to hear more feedback so that we decide next steps.

Reply:
This ties in to my previous comments to @halter73 about overhauling DI integration. The fact that calling WithToolXx multiple times right now leads to handlers being overwritten is problematic, and the design needs to accomodate that. That relates to your suggestion of chaining.  Can you open an issue?